router: do not reset started response if per_try_timeout occured

curlup opened the pull request:
Signed-off-by: Pavel Trukhanov <pavel.trukhanov@gmail.com>

*Description*: We stumbled upon a very strange behaviour wrt per_try_timeout. We have a service behind envoy. The "usual" serving time is, say, 40ms, so we configured per_try_timeout to 50ms and global to 100ms. We've started to see interruptions in service from the client side - sometimes client receives "200 ok" and some headers but fails on `EOF` while reading response body. And timings were strange — ~50ms. 
As I dug the code, I found [this part](https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L491) stating
``` 
// If we have not yet sent anything downstream, send a response with an appropriate status code.
// Otherwise just reset the ongoing response.
```
Which is getting executed in case of Global/ResponseTimeout, as well as PerTryTimeout. 

As I understand that's made to respect global timeout guarantees for clients.
Nevertheless in my understanding that's undesirable behaviour in case of PerTryTimeout firing up while there's still time until global timeout. Because otherwise envoy effectively produces errors for the client (EOFs by reset), while all we wanted by configuring per_try_timeout was to run a retry if upstream didn't respond in a timely manner.
So I think the proper solution (which I tried to implement here) is not to cancel and reset response, if per_try_timeout was fired and we still have time until global timeout. And just ignore fired per_try_timeout, if upstream have already started response and we've sent some of it downstream.



*Risk Level*: Low | Medium

*Testing*: I wasn't able to run tests (and fix what is needed) because ci docker helper refuses to download something from GH. I'm hoping on ci to run them, and then I'd be happy to fix/write more tests etc.

*Docs Changes*: I didn't find this peculiar behaviour to be described in docs (it might be I wasn't looking thoroughly enough). So I don't know if it's needed to be put there at all.

*Release Notes*: N/A (correct me if I'm wrong)

mattklein123 commented:
@curlup We've run into this internally at Lyft, and I think arguments can be made both for and against doing what you propose (I think I could be swayed to change the behavior). Let me think about it and I will get back to you.

mattklein123 commented:
@curlup sorry for the delay. I thought about it, and I think this change makes sense and we should move forward with it. Will you have time to add tests, release notes, etc.?

curlup commented:
Yes, thanks!

What kind of tests do you want? 
CI seems to be ok with the current set. 

Regarding docs — what's the best place it should be documented? 

Should a release note just state that a bug in particular behaviour was fixed or do you think if something else?

mattklein123 commented:
@curlup I would just do a small release note explaining the change in behavior as it's not obvious. 

For tests, take a look at the tests here: https://github.com/envoyproxy/envoy/blob/master/test/common/router/router_test.cc#L977. I would just add a new test in which we get the per-try timeout after we already started to proxy the response. Your change in behavior should cause that test to fail with the old code and then it can be fixed to support your new changes. I don't think it will be that difficult

curlup commented:
K, will do

curlup commented:
check #3344 for test - it's failing on master, while passes here (cherry-picked commit) due to fix.

Regarding unsuccessful CI run - I don't know why on darwin metrics_service_integration_test test has failed. Seems random. Any ideas?

mattklein123 commented on test/common/router/router_test.cc:
```diff
@@ -1126,6 +1126,36 @@ TEST_F(RouterTest, RetryUpstreamPerTryTimeout) {
+  EXPECT_CALL(cm_.conn_pool_.host_->outlier_detector_, putHttpResponseCode(200));
+  response_decoder->decodeHeaders(std::move(response_headers), false);
+  per_try_timeout_->callback_();
+  router_.onDestroy();
```
If you want to remove the onDestroy(), finish the response from upstream (basically dispatch data with end_stream = true).

curlup replied:
Ok, thank. Now I understand. Do you want me to do that though, or Destroy is fine like that?

curlup replied:
Got it. Done and tested

mattklein123 commented:
Also, please add a small release note on the behavior change.

curlup commented:
I've added Release Note earlier:
Release Notes: Fixed a bug - unneeded request canceling in case it's already started but per_try_timeout kicks in.

If you need a longer one: now Envoy won't cancel and reset an ongoing response in case of per_try_timeout but will allow it to (try to) finish up until global timeout.

curlup commented:
added release note as well

mattklein123 commented on docs/root/intro/version_history.rst:
```diff
@@ -61,6 +61,8 @@ Version history
 * logger: added the ability to optionally set the log format via the :option:`--log-format` option.
 * logger: all :ref:`logging levels <operations_admin_interface_logging>` can be configured
   at run-time: trace debug info warning error critical.
+* router: fixed a bug - unneeded response canceling on per_try_timeout in case response has already started, 
```
nit: Can you use the following phrasing:

```
router: The behavior of per-try timeouts have changed in the case where a portion of the response has 
already been proxied downstream when the timeout occurs. Previously, the response would be reset 
leading to either an HTTP/2 reset or an HTTP/1 closed connection and a partial response. Now, the 
timeout will be ignored and the response will continue to proxy up to the global request timeout. 
```

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -928,16 +928,25 @@ void Filter::UpstreamRequest::setupPerTryTimeout() {
-  parent_.onUpstreamReset(
-      UpstreamResetType::PerTryTimeout,
-      absl::optional<Http::StreamResetReason>(Http::StreamResetReason::LocalReset));
+  // If we've sent anything downstream but per try timeout occurred
```
nit: Can you reflow this comment to 100 cols. Also, rephrase:
"If we've sent anything downstream, ignore the per try timeout and let the response continue up to the global timeout."

mattklein123 commented on test/common/router/router_test.cc:
```diff
@@ -1201,6 +1201,39 @@ TEST_F(RouterTest, RetryUpstreamPerTryTimeout) {
+  EXPECT_CALL(cm_.conn_pool_.host_->outlier_detector_, putHttpResponseCode(200));
+  response_decoder->decodeHeaders(std::move(response_headers), false);
+  per_try_timeout_->callback_();
+  response_decoder->decodeData(body, true);
```
nit: Can you add something like `EXPECT_CALL(callbacks_, encodeData(_, true));` above this line? We should expect that we proxy the data.

curlup commented:
I did all you've asked. At least I think so =)

ipv6_tests is somewhat flaky I think - it passed just now, on the previous commit, the one before version_history.rst conflict resolving commit.
